### PR TITLE
Label shift

### PR DESCRIPTION
**Pull request by TammoR**
This implements the black box label-shift predictor as described [here](https://arxiv.org/abs/1802.03916) and also implements empirical risk minimization to fix a shift if it is detected. The important assumption here is that we have a pure label shift in the data, while the label conditionals are constant that is p(covariates|label) = q(covariates|label), where p and q are the training and production distributions, respectively.

## Shift detection
The API can be best seen in ```test_simple_imputer_label_shift()``` inside ```test_simple_imputer.py```. 
Different from previous discussions, we don't need the test or training data to detect a label shift, because confusion matrices and marginals are sufficient and already stored as part of the imputer metrics.

Thus, after a simple imputer is trained we can now simply call:
``` 
weights = imputer.check_for_label_shift(production_data) 
```

This will a return a dictionary with labels as keys and weights for ERM as values. It will also sent a message to the logger as follows:

```
The estimated true label marginals are [('label_0', 0.62), ('label_1', 0.38)]
Marginals in the training data are [('label_0', 0.23), ('label_1', 0.77)]
Reweighing factors for empirical risk minimization{'label_0': 2.72, 'label_1': 0.49}
The smallest eigenvalue of the confusion matrix is 0.21 ' (needs to be > 0).
```

## Empirical risk minimization and fixing label shift
To perform empirical risk minimization we need to retrain the model with the computed class weights. This is done by passing the weights dictionary as command line argument to the ```.fit()``` method of the SimpleImputer.

```
simple_imputer.fit(train_df, class_weights)
```

Training the imputation model performs empirical risk minimization by weighing every observation's contribution to the log-likelihood. Alternatively we can pass a list with weights for every single instance of the training data, e.g. if we believe that particular observations are more reliable

```
simple_imputer.fit(train_df, instance_weights)
```

### CODE REVIEW

File: datawig/utils.py [hunk 1]
```diff
@@ -395,3 +397,58 @@ def sample_cartesian(sets: List,
...
+    :param weights: dictionary with label names and corresponding weights
+    :param label_column: name of label column
+    """
+
+    if in_place is False:
+        df_new = df.copy()
+    else:
+        df_new = df
+
+    df_new['__empirical_risk_instance_weight__'] = 0.0
```
Comment (tdhd) [inline datawig/utils.py]: also use the variable for `__empirical_risk_instance_weight__` here please

File: datawig/_hpo.py [hunk 2]
```diff
@@ -73,16 +73,16 @@ def __preprocess_hps(self,
...
+        default_hps['global']['weight_decay'] = [0]
         default_hps['global']['num_epochs'] = [25]
         default_hps['global']['patience'] = [5]
         default_hps['global']['batch_size'] = [16]
         default_hps['global']['final_fc_hidden_units'] = [[]]
         default_hps['string'] = {}
         default_hps['string']['ngram_range'] = {}
         default_hps['string']['max_tokens'] = [2 ** 15]
-        default_hps['string']['tokens'] = [['words']]
+        default_hps['string']['tokens'] = [['chars']]
```
Comment (tdhd) [inline datawig/_hpo.py]: This shouldn't have changed, is your branch up to date with master?

File: datawig/simple_imputer.py [hunk 3]
```diff
@@ -27,11 +27,12 @@
 import pandas as pd
 import glob
 import shutil
+import numpy as np
 from pandas.api.types import is_numeric_dtype
-from sklearn.metrics import mean_squared_error, f1_score, precision_score, accuracy_score, recall_score
+from sklearn.metrics import confusion_matrix
```
Comment (tdhd) [inline datawig/simple_imputer.py]: is this import used?

File: datawig/utils.py [hunk 4]
```diff
@@ -395,3 +397,58 @@ def sample_cartesian(sets: List,
     assert width == 1  # at the end of this procedure, the leaf index set should have width 1.
 
     return out
+
+
+### Functionality for label shift detection. Experimental. ###
+
+def synthetic_label_shift_simple(N, label_proportions, error_proba, covariates=None) -> pd.DataFrame:
```
Comment (tdhd) [inline datawig/utils.py]: This function is only used in tests I think, please move it to a test module, conftest.py could be a good place for this.

**Comment by tdhd** [comment]
Thanks for the PR Tammo!

File: datawig/_hpo.py [hunk 5]
```diff
@@ -198,12 +198,13 @@ def my_function(true, predicted, confidence):
...
                     # call kw. args. with: **{key: item for key, item in col_parms.items() if not key == 'type'})]
-                    data_encoders += [TfIdfEncoder(input_columns=[input_column],
-                                                   output_column=input_column + '_' + token,
-                                                   tokens=token,
-                                                   ngram_range=col_parms['ngram_range:'+token],
-                                                   max_tokens=col_parms['max_tokens'])]
+                    data_encoders += [encoder(input_columns=[input_column],
+                                              output_column=input_column + '_' + token,
+                                              tokens=token,
+                                              ngram_range=col_parms['ngram_range:'+token],
```
Comment (sscdotopen) [inline datawig/_hpo.py]: missing spaces around +

File: datawig/simple_imputer.py [hunk 6]
```diff
@@ -320,8 +325,22 @@ def fit(self,
...
+        :param instance_weights: List of weights for each instance in train_df.
         """
+
+        # add weights to training data if provided
+        if class_weights is not None:
+            assert instance_weights is None
+            train_df = add_class_weight_to_df(train_df, class_weights, self.output_column, in_place=False)
+
+        elif instance_weights is not None:
+            assert class_weights is None
```
Comment (sscdotopen) [inline datawig/simple_imputer.py]: can be removed because it has been checked in line 335

File: datawig/simple_imputer.py [hunk 7]
```diff
@@ -320,8 +325,22 @@ def fit(self,
...
+            train_df = train_df.copy()
```
Comment (sscdotopen) [inline datawig/simple_imputer.py]: minor comment: maybe it would be nicer to wrap the input data with an iterator that adds the instance weights on the fly instead of forcing a full copy of the input data.

File: datawig/simple_imputer.py [hunk 8]
```diff
@@ -586,3 +603,65 @@ def load_hpo_model(self, hpo_name: int = None):
...
+        # generate predictions for test and target data
+        imputed_target = self.predict(target_data)
+        # imputed_test = self.predict(test_data)
+
+        # invert normalized empirical confusion matrix for test data
+        ids = dict((label, index) for (index, label) in enumerate(labels))
+        confusion_matrix_from_file = self.load_metrics()['confusion_matrix']
+        confusion_test = np.empty([len(confusion_matrix_from_file), len(confusion_matrix_from_file)])
+
+        for row in confusion_matrix_from_file:
```
Comment (sscdotopen) [inline datawig/simple_imputer.py]: maybe put this in a helper method and comment it, its a bit confusing to read (probably only because I dont know the serialization format of the confusion matrix)

File: datawig/simple_imputer.py [hunk 9]
```diff
@@ -586,3 +603,65 @@ def load_hpo_model(self, hpo_name: int = None):
...
+                        col_idx = ids[col[0]]
+                        confusion_test[row_idx, col_idx] = col[1]
+
+        confusion_test = confusion_test/confusion_test.sum()
+        confusion_inv_test = np.linalg.inv(confusion_test)
+
+        # compute estimates of label marginals for test and target data
+        # marginals_test = imputed_test[self.output_column+'_imputed'].value_counts(normalize=True, sort=False)[labels]
+        marginals_test = pd.Series(confusion_test.sum(axis=0), index=labels)
+        marginals_target = imputed_target[self.output_column+'_imputed'].value_counts(
```
Comment (sscdotopen) [inline datawig/simple_imputer.py]: missing spaces around +

File: datawig/simple_imputer.py [hunk 10]
```diff
@@ -586,3 +603,65 @@ def load_hpo_model(self, hpo_name: int = None):
...
+            normalize=True, sort=False)[labels]
+
+        # estimate the ratio of marginals and store as dictionary.
+        label_weights = confusion_inv_test.dot(marginals_target)
```
Comment (sscdotopen) [inline datawig/simple_imputer.py]: why are we computing the inverse and not solving a linear system via QR or so?

File: datawig/utils.py [hunk 11]
```diff
@@ -395,3 +397,58 @@ def sample_cartesian(sets: List,
...
+
+    out = []
+    for n in range(N):
+        label = np.random.choice(range(len(label_proportions)), p=label_proportions)
+        if np.random.rand() > error_proba:
+            covariate = covariates[label]
+        else:
+            # choose a different covariate at random.
+            covariate = covariates[np.random.choice([i for i in range(len(label_proportions)) if i != label])]
+        out.append((covariate, 'label_'+str(label)))
```
Comment (sscdotopen) [inline datawig/utils.py]: missing spaces around +

File: datawig/utils.py [hunk 12]
```diff
@@ -395,3 +397,58 @@ def sample_cartesian(sets: List,
...
+    :param N: Number of observations
+    :param label_proportions: Dirichlet vector with label proportions
+    :param error_proba: Probability of sampling from a different covariate distribution
+    :param covariates: list of covariate names, random strings by default
+    """
+
+    if covariates is None:
+        covariates = []
+        for i in range(len(label_proportions)):
+            covariates.append(''.join([random.choice(string.ascii_letters.lower()) for n in range(6)]))
```
Comment (felixbiessmann) [inline datawig/utils.py]: There's also a [`rand_str`](https://github.com/awslabs/datawig/blob/23deab456060bc393224002211207d0c9b0b915e/datawig/utils.py#L135) function for that in utils.

File: datawig/simple_imputer.py [hunk 13]
```diff
@@ -320,8 +325,22 @@ def fit(self,
...
+
+        # add weights to training data if provided
+        if class_weights is not None:
+            assert instance_weights is None
+            train_df = add_class_weight_to_df(train_df, class_weights, self.output_column, in_place=False)
+
+        elif instance_weights is not None:
+            assert class_weights is None
+            train_df = train_df.copy()
+            train_df['__empirical_risk_instance_weight__'] = instance_weights
```
[the remark this answers is not in the document]
Comment (TammoR) [reply datawig/simple_imputer.py]: Good point!

File: datawig/simple_imputer.py (hunk 10 above)
Comment (TammoR) [reply datawig/simple_imputer.py]: Thanks for catching that. We were interested in the inverse at some point, but now it's much better to use something more stable, PLU decomposition here. (np.linalg.solve)

File: datawig/utils.py (hunk 12 above)
Comment (TammoR) [reply datawig/utils.py]: good call, thanks.

File: datawig/simple_imputer.py [hunk 14]
```diff
@@ -588,3 +595,107 @@ def load_hpo_model(self, hpo_name: int = None):
...
+        logger.warn('\n\tThe estimated label marginals are ' + str(list(zip(labels, true_marginals_target))) +
+                    '\n\tMarginals in the training data are ' + str(list(zip(labels, marginals_test))) +
+                    '\n\tReweighing factors for empirical risk minimization' + str(label_weights_dict))
+
+        if np.any(marginals_test < 0):
+            logger.warn('\n\tEstimated label marginals are invalid. Proceed with caution.')
+
+        return label_weights_dict
+
+    def add_weights_to_df(self,
```
Comment (tdhd) [inline datawig/simple_imputer.py]: Please change this to a "private" method `def __add_weights_to_df(...)`.
